indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskRunner.java

```diff
@@ -88,11 +89,19 @@
    */
   void shutdown(String taskid, String reason);
 
-  default void shutdown(String taskid, String reasonFormat, Object... args)
+  default void shutdown(String taskid, String reasonFormat, Object... reasonArgs)
   {
-    shutdown(taskid, StringUtils.format(reasonFormat, args));
+    // only calculate the 'reason' string for debug level logging
+    // in large clusters the 'reasonArgs' may be very large / expensive if it includes a list of tasks
+    String reason = (getLogger().isDebugEnabled()) ? StringUtils.format(reasonFormat, reasonArgs) : "debug log disabled";
```

Comment:
If debug logging is not enabled, then the following log line won't be reported. 
https://github.com/apache/druid/blob/master/indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java#L582

Reply:
Similarly, this - 
https://github.com/apache/druid/blob/master/indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java#L299

Reply:
Looking at the callers who pass in taskIds, I see only TaskQueue#shutdown() that is passing in the task ids. So, I propose we modify code in TaskQueue#shutdown something like this:
```
// Kill tasks that shouldn't be running
    final Set<String> knownTaskIds = tasks
        .stream()
        .map(Task::getId)
        .collect(Collectors.toSet());
    final Set<String> tasksToKill = Sets.difference(runnerTaskFutures.keySet(), knownTaskIds);
    if (!tasksToKill.isEmpty()) {
      log.info("Asking taskRunner to clean up %,d tasks.", tasksToKill.size());
      // On large installations running several thousands of tasks,
      // concatenating the list of known task ids can be compupationally expensive.
      boolean logKnownTaskIds = log.isDebugEnabled();
      String reason = logKnownTaskIds
                      ? String.format("Task is not in knownTaskIds[%s]", knownTaskIds)
                      : "Task is not in knownTaskIds";
      for (final String taskId : tasksToKill) {
        try {
          taskRunner.shutdown(
              taskId, reason
          );
        }
        catch (Exception e) {
          log.warn(e, "TaskRunner failed to clean up task: %s", taskId);
        }
      }
    }
```

If we do this, then you won't have to add a new `getLogger()` method in `TaskRunner` interface.

Reply:
I like this solution, I'll swap over to this implementation.

Reply:
thanks @samarthjain , this solution is much simpler - fixes applied

let me know if all good and i can squash it